lib/eth/tx.rb

```diff
@@ -70,19 +78,28 @@ def to_h
     end
 
     def from
-      if signature
-        public_key = OpenSsl.recover_compact(signature_hash, signature)
+      if ecdsa_signature
+        public_key = OpenSsl.recover_compact(signature_hash, ecdsa_signature)
         Utils.public_key_to_address(public_key) if public_key
       end
     end
 
     def signature
       return @signature if @signature
-      self.signature = [
-        Utils.int_to_base256(v),
-        Utils.zpad_int(r),
-        Utils.zpad_int(s),
-      ].join if [v, r, s].all?
+      @signature = { v: v, r: r, s: s } if [v, r, s].all? && (v > 0)
+    end
+
+    def ecdsa_signature
```

Comment:
The signature method maybe should be renamed eip155_signature (but maybe not, because if chain_id is nil, then the signature is replayable). The two methods reflect the two uses of the signature: ecdsa_signature always has v == 27 or 28, and is used in the open_ssl calls to avoid the overflow bug. The signature method returns the signature that would be packed into the transaction, and includes the chain_id adjustment. Note that ecdsa_signature returns a packed value, because v is 27 or 28, and because open_ssl expects that format; signature, on the other hand, returns a hash where v is an integer value and we don't have to worry about overflowing.

So I think the answer is: we probably shouldn't consolidate them, because they really address different uses. There may be a way to adjust prior to the open_ssl calls, but I thought that may be klugier that having the two methods.